# Conceptual understanding of basecalling
*This aims to show an example that can aid in simplifying the complexity in DNA sequencing using ONTs nanopore devices. In general, we can view the processing done by the ONT devices as encoding from organic DNA to an electrical signal. However, the encoded output from ONTs devices is impossible for humans to interpret (decode) due to the complex signal generated by the nanopore. Interpreting the signal is referred to as basecalling and is done using artificial intelligence. The topic of this is not to make any statements regarding how to process the signal. The sole purpose is to define tangible examples that can be used to understand the sequencing process itself and underlying obstacles regarding working with the DNA material.*

We see multiple advantages of such example:
1. Explaining the problem to inexperienced people becomes easier
2. Discussions can take a basis in more tangible examples
3. Uncovering differences between basecalling and the simplified example can provoke realizations about misunderstandings
4. (In)validate whether the `Long term dependencies NLP`-analogy holds for this domain*

*Long term dependencies in NLP means that correlations between words are not only depended on the distance between them 

## DNA-signal creation
*We explain the process of creating the signal while avoiding to use biological- and machine intelligence-specific terms. This is done to understand the correlation to the example, we are later introducing.*

A DNA sequence can loosely be seen as a sequence over the alphabet `ACGT`. 

Example: 
> AACGTGTGACA

In the process of converting the sequence of the DNA, an integer value is created that captures the electrical resistance of 5 bases. Five such bases are referred to as a 5-mer.

Example: 
> AACGT => 354

The sequencing device captures these values as the DNA moves through. The DNA does not move at a fixed speed, but in general, we can simplify the problem by saying, that a 5-mer is captured by somewhere between 5-10 reads. 


### Properties of the sequence
The encoded signal produced by the device is, therefore:
1. A sequence of integers defining the resistances of contexts
2. The signal for context $c_i$ is not affected by the signal of context $c_{i+y}$ where $y$ can be any given value.

In addition we will argue, that the following also holds:

3. The change between signal $s_i$ and $s_{i+1}$ provide *some information* about the change of the context:
    - The same of course goes for $s_{i+1}$ and $s_{i+2}$
    - **This means that:** The change between $s_i$ and $s_{i+2}$ also provide some information about the change in context
    - **But:** reads close to each other have higher correlation (as we will argue in our example)


## Questions
The signal output range from the MinION is somewhere between 200 and 850 yielding 650 possible values. DNA consists of 4 different bases of which may have 1 (or 2?) modifications, that affect their signal. Assuming that the nanopore always reads 5 (and only 5) bases, then this produces $5^5$ (or $5^6$) different combinations of contexts. In an ideal scenario, we would be able to make a function that takes a signal read and returns the correct context. This ideal scenario is not possible since it is impossible to create a function where the number of inputs (650) is lower than the number of outputs ($5^5$ or $5^6$).

This is just a testament to the complexity of the task of basecalling. To make valid arguments about a solution, we need a better understanding of the correlations between input and output of the MinION device.

- Does the data that we are provided include modified bases, or is that only done artificially?
- Does any combination of 5 bases produce the same read? - (or are bases closer to the middle more dominating?)
  - e.g. are these signals equal: ACGAA = AAACG
- If not, does the reverse of a 5-mer produce the same read?
  - e.g. are these signals equal: ACGAA = AAGCA
- Is there an instant switch between two contexts - or is it a continuous transition? See below:

  **IS THE FOLLOWING CORRECT?**
  In reality, the device will not measure a set of reads for a 5-mer and then the following read will be of a new 5-mer (e.i one base moved out and a new moved in). Instead, are the first couple of reads of a 5-mer also influenced by the 5-mer before it and the last reads are influenced by the 5-mer after.

  Example: 
  ```
  aACGTG  => 350
  aACGTG  => 350
  ACGTG  => 355
  ACGTG  => 355
  ACGTGt => 360
  ACGTGt => 360
  ```
  Where lowercase letters are only partial influencing the signal. Everything influencing a given read is refered to by the *context*, whereas a *5-mer* is the five capital letters.



## Example
Imagine the hypothetic DNA, that consists of the bases `A`, `B`, and `C` and a device that only captures 2-mers. Assume that the device is not influenced by noise and makes one read per 2-mer. These adjustments greatly simplify the problem, but every *property of the sequence* still holds.

Giving these assumptions we can make a table with the signal of every 2-mer (2^3 = 9 different possibilities).

| 2-mer  | AA | AB | AC | BA | BB | BC | CA | CB | CC |
|--------|----|----|----|----|----|----|----|----|----|
|**Read**| 1  | 1  | 1  | 3  | 2  | 3  | 4  | 4  | 2  |

If given the signal value 1, we can make a lookup in the table, and see that this either corresponds to AA, AB or AC. Only provided with a single read does not enable us to deterministically select the correct 2-mer - but we have narrowed it down. If the next read is also a 1, we know that the first 2-mer is AA, because the only way two reads of 1 can overlap, is if both 2-mers are AA. If the second read instead is 2, it means that first 2-mer is either AB or AC. We can use this process of overlapping k-mers until the correct initial one is determined (given that the values for the different k-mers are well distributed). The following shows the graph of how the sequence [1, 2, 4] deterministically chooses the 2-mer AC. 

<img src="@attachment/2mer.png">

It is obvious, that we cannot make an overlap if we skip over a 2-mer, since the two 2-mers will not share any bases. Extending the example to 3-mers (or higher) will show that skipping will still result in un-determinism. Take the example, that we are choosing whether the 3-mer is ABA or ACA. If we skip the read of the second 3-mer we can only make an overlap on the third base. Regardless of what 3-mer starting with A that the third read suggests, we would not be able to decide between ABA or ACA. Whereas an overlap from the second read would aid with that.

By this we can conclude when determining any k-mer $i$, that signal $S_{i+k}$ has more importance than signal $S_{i+k+1}$.

